be/src/storage/lake/lake_persistent_index.cpp

```diff
@@ -175,8 +218,114 @@ Status LakePersistentIndex::try_replace(size_t n, const Slice* keys, const Index
+    while (iter_ptr->Valid()) {
+        RETURN_IF_ERROR(merger->merge(iter_ptr->key().to_string(), iter_ptr->value().to_string()));
+        iter_ptr->Next();
+    }
+    RETURN_IF_ERROR(iter_ptr->status());
+    merger->finish();
+    return builder->Finish();
+}
+
+Status LakePersistentIndex::major_compact(int64_t min_retain_version, std::shared_ptr<TxnLogPB>& txn_log) {
+    if (_sstables.size() < 2) {
+        return Status::OK();
+    }
+
+    auto iter_ptr = prepare_merging_iterator();
+    if (!iter_ptr->Valid()) {
+        return iter_ptr->status();
+    }
+
+    auto filename = gen_sst_filename();
+    auto location = _tablet_mgr->sst_location(_tablet_id, filename);
+    ASSIGN_OR_RETURN(auto wf, fs::new_writable_file(location));
+    sstable::Options options;
+    std::unique_ptr<sstable::FilterPolicy> filter_policy;
+    filter_policy.reset(const_cast<sstable::FilterPolicy*>(sstable::NewBloomFilterPolicy(10)));
+    options.filter_policy = filter_policy.get();
+    sstable::TableBuilder builder(options, wf.get());
+    RETURN_IF_ERROR(merge_sstables(std::move(iter_ptr), &builder));
+    RETURN_IF_ERROR(wf->close());
+
+    auto max_compaction_versions = config::lake_pk_index_sst_max_compaction_versions;
+    for (const auto& sstable : _sstables) {
+        auto input_sstable = txn_log->mutable_op_compaction()->add_input_sstables();
+        auto sstable_pb = sstable->sstable_pb();
+        input_sstable->CopyFrom(sstable_pb);
+        if (txn_log->op_compaction().input_sstables_size() >= max_compaction_versions) {
+            break;
+        }
+    }
+    txn_log->mutable_op_compaction()->mutable_output_sstable()->set_filename(filename);
+    txn_log->mutable_op_compaction()->mutable_output_sstable()->set_filesize(builder.FileSize());
+    return Status::OK();
+}
+
+Status LakePersistentIndex::apply_opcompaction(const TxnLogPB_OpCompaction& op_compaction) {
+    if (op_compaction.input_sstables().empty()) {
+        return Status::OK();
+    }
+
+    PersistentIndexSstablePB sstable_pb;
+    sstable_pb.CopyFrom(op_compaction.output_sstable());
+    sstable_pb.set_version(op_compaction.input_sstables(op_compaction.input_sstables().size() - 1).version());
+    auto sstable = std::make_unique<PersistentIndexSstable>();
+    RandomAccessFileOptions opts{.skip_fill_local_cache = true};
+    ASSIGN_OR_RETURN(auto rf,
+                     fs::new_random_access_file(opts, _tablet_mgr->sst_location(_tablet_id, sstable_pb.filename())));
+    RETURN_IF_ERROR(sstable->init(std::move(rf), sstable_pb, nullptr));
+
+    std::unordered_set<std::string> filenames;
+    for (const auto& input_sstable : op_compaction.input_sstables()) {
+        filenames.insert(input_sstable.filename());
+    }
+    auto it = _sstables.begin();
+    while (it != _sstables.end()) {
+        auto sstable_pb = (*it)->sstable_pb();
+        if (filenames.contains(sstable_pb.filename())) {
+            it = _sstables.erase(it);
```

Comment:
vector::erase is linear complexity plus the move of the elements

suggest replace it with `std::remove_if` and `std::vector::erase`.